osquery/core/query.cpp

```diff
@@ -468,240 +329,195 @@ DiffResults diff(const QueryData& old, const QueryData& current) {
-    auto status = serializeQueryData(item.snapshot_results, snapshot_results);
+    auto arr = doc.getArray();
+    auto status = serializeQueryData(item.snapshot_results, temp_doc, arr);
     if (!status.ok()) {
       return status;
     }
-    results.add_child("snapshot", snapshot_results);
+    temp_doc.add("snapshot", arr);
   } else {
     // This error case may also be represented in serializeQueryLogItem.
     return Status(1, "No diff results or snapshot results");
   }
 
-  for (const auto& action : results) {
-    for (const auto& row : action.second) {
-      pt::ptree event;
-      serializeEvent(item, row.second, event);
-      event.put<std::string>("action", action.first);
-      tree.push_back(std::make_pair("", event));
+  for (auto& action : temp_doc.doc().GetObject()) {
+    for (auto& row : action.value.GetArray()) {
+      auto obj = doc.getObject();
+      serializeEvent(item, row, doc, obj);
+      doc.addCopy("action", action.name.GetString(), obj);
+      doc.push(obj);
     }
   }
-  return Status(0, "OK");
+  return Status();
 }
 
 Status serializeQueryLogItemJSON(const QueryLogItem& i, std::string& json) {
-  pt::ptree tree;
-  auto status = serializeQueryLogItem(i, tree);
+  auto doc = JSON::newObject();
+  auto status = serializeQueryLogItem(i, doc);
   if (!status.ok()) {
     return status;
   }
 
-  std::ostringstream output;
-  try {
-    pt::write_json(output, tree, false);
-  } catch (const pt::json_parser::json_parser_error& e) {
-    // The content could not be represented as JSON.
-    return Status(1, e.what());
-  }
-  json = output.str();
-  return Status(0, "OK");
+  return doc.toString(json);
 }
 
-Status deserializeQueryLogItem(const pt::ptree& tree, QueryLogItem& item) {
-  if (tree.count("diffResults") > 0) {
-    auto status =
-        deserializeDiffResults(tree.get_child("diffResults"), item.results);
+Status deserializeQueryLogItem(const rj::Document& doc, QueryLogItem& item) {
+  if (doc.HasMember("diffResults")) {
+    auto status = deserializeDiffResults(doc["diffResults"], item.results);
     if (!status.ok()) {
       return status;
     }
-  } else if (tree.count("snapshot") > 0) {
-    auto status =
-        deserializeQueryData(tree.get_child("snapshot"), item.snapshot_results);
+  } else if (doc.HasMember("snapshot")) {
+    auto status = deserializeQueryData(doc["snapshot"], item.snapshot_results);
     if (!status.ok()) {
       return status;
     }
   }
 
-  getLegacyFieldsAndDecorations(tree, item);
-  return Status(0, "OK");
+  getLegacyFieldsAndDecorations(doc, item);
+  return Status();
 }
 
 Status deserializeQueryLogItemJSON(const std::string& json,
                                    QueryLogItem& item) {
-  pt::ptree tree;
-  try {
-    std::stringstream input;
-    input << json;
-    pt::read_json(input, tree);
-  } catch (const pt::json_parser::json_parser_error& e) {
-    return Status(1, e.what());
+  rj::Document doc;
+  if (doc.Parse(json.c_str()).HasParseError()) {
```

Comment:
When Parse() accepts a `std::string` it automatically converts it to a C-Style string.
I suggest not converting it explicitly to allow RapidJSON to handle the string as it sees fit.
See https://github.com/Tencent/rapidjson/blob/master/include/rapidjson/document.h#L2346

Reply:
I'm also wondering if we can destruct the original JSON string.
It makes parsing much faster as far as I recall.